// mkdist/dist/shared/mkdist.72f9e2c9.cjs
'use strict';

const pathe = require('pathe');
const fse = require('fs-extra');
const node_stream = require('node:stream');
const node_fs = require('node:fs');
const esbuild = require('esbuild');
const jiti = require('jiti');
const node_url = require('node:url');
const mlly = require('mlly');

function copyFileWithStream(sourcePath, outPath) {
  const sourceStream = node_fs.createReadStream(sourcePath);
  const outStream = node_fs.createWriteStream(outPath);
  return new Promise((resolve, reject) => {
    node_stream.pipeline(sourceStream, outStream, (error) => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

const DECLARATION_RE = /\.d\.[cm]?ts$/;
const CM_LETTER_RE = /(?<=\.)(c|m)(?=[jt]s$)/;
const jsLoader = async (input, { options }) => {
  if (![".ts", ".js", ".cjs", ".mjs"].includes(input.extension) || DECLARATION_RE.test(input.path)) {
    return;
  }
  const output = [];
  let contents = await input.getContents();
  if (options.declaration && !input.srcPath?.match(DECLARATION_RE)) {
    const cm = input.srcPath?.match(CM_LETTER_RE)?.[0] || "";
    const extension2 = `.d.${cm}ts`;
    output.push({
      contents,
      srcPath: input.srcPath,
      path: input.path,
      extension: extension2,
      declaration: true
    });
  }
  if (input.extension === ".ts") {
    contents = await esbuild.transform(contents, { loader: "ts" }).then((r) => r.code);
  }
  const isCjs = options.format === "cjs";
  if (isCjs) {
    contents = jiti().transform({ source: contents, retainLines: false }).replace(/^exports.default = /gm, "module.exports = ");
  }
  let extension = isCjs ? ".js" : ".mjs";
  if (options.ext) {
    extension = options.ext.startsWith(".") ? options.ext : `.${options.ext}`;
  }
  output.push({
    contents,
    path: input.path,
    extension
  });
  return output;
};

const sassLoader$1 = async (input) => {
  if (![".sass", ".scss"].includes(input.extension)) {
    return;
  }
  if (pathe.basename(input.srcPath).startsWith("_")) {
    return [
      {
        contents: "",
        path: input.path,
        skip: true
      }
    ];
  }
  const compileString = await import('sass').then(
    (r) => r.compileString || r.default.compileString
  );
  const output = [];
  const contents = await input.getContents();
  output.push({
    contents: compileString(contents, {
      loadPaths: ["node_modules"],
      url: node_url.pathToFileURL(input.srcPath)
    }).css,
    path: input.path,
    extension: ".css"
  });
  return output;
};

const vueLoader = async (input, context) => {
  if (input.extension !== ".vue") {
    return;
  }
  const output = [
    {
      path: input.path,
      contents: await input.getContents()
    }
  ];
  let earlyReturn = true;
  for (const blockLoader of [sassLoader, scriptLoader]) {
    const result = await blockLoader(
      { ...input, getContents: () => output[0].contents },
      context
    );
    if (!result) {
      continue;
    }
    earlyReturn = false;
    const [vueFile, ...files] = result;
    output[0] = vueFile;
    output.push(...files);
  }
  if (earlyReturn) {
    return;
  }
  return output;
};
const vueBlockLoader = (options) => async (input, { loadFile }) => {
  const contents = await input.getContents();
  const BLOCK_RE = new RegExp(
    `<${options.type}((\\s[^>\\s]*)*)>([\\S\\s.]*?)<\\/${options.type}>`
  );
  const [block, attributes = "", _, blockContents] = contents.match(BLOCK_RE) || [];
  if (!block || !blockContents) {
    return;
  }
  if (options.exclude?.some((re) => re.test(attributes))) {
    return;
  }
  const [, lang = options.outputLang] = attributes.match(/lang="([a-z]*)"/) || [];
  const extension = "." + lang;
  const files = await loadFile({
    getContents: () => blockContents,
    path: `${input.path}${extension}`,
    srcPath: `${input.srcPath}${extension}`,
    extension
  }) || [];
  const blockOutputFile = files.find(
    (f) => f.extension === `.${options.outputLang}` || options.validExtensions?.includes(f.extension)
  );
  if (!blockOutputFile) {
    return;
  }
  const newAttributes = attributes.replace(
    new RegExp(`\\s?lang="${lang}"`),
    ""
  );
  return [
    {
      path: input.path,
      contents: contents.replace(
        block,
        `<${options.type}${newAttributes}>
${blockOutputFile.contents?.trim()}
</${options.type}>`
      )
    },
    ...files.filter((f) => f !== blockOutputFile)
  ];
};
const sassLoader = vueBlockLoader({
  outputLang: "css",
  type: "style"
});
const scriptLoader = vueBlockLoader({
  outputLang: "js",
  type: "script",
  exclude: [/\bsetup\b/],
  validExtensions: [".js", ".mjs"]
});

const defaultLoaders = [vueLoader, jsLoader, sassLoader$1];
function createLoader(loaderOptions = {}) {
  const loaders = loaderOptions.loaders || defaultLoaders;
  const loadFile = async function(input) {
    const context = {
      loadFile,
      options: loaderOptions
    };
    for (const loader of loaders) {
      const outputs = await loader(input, context);
      if (outputs?.length) {
        return outputs;
      }
    }
    return [
      {
        path: input.path,
        srcPath: input.srcPath,
        raw: true
      }
    ];
  };
  return {
    loadFile
  };
}

async function getDeclarations(vfs, opts) {
  const ts = await import('typescript').then((r) => r.default || r);
  const inputFiles = [...vfs.keys()];
  const compilerOptions = {
    allowJs: true,
    declaration: true,
    incremental: true,
    skipLibCheck: true,
    strictNullChecks: true,
    emitDeclarationOnly: true
  };
  const tsHost = ts.createCompilerHost(compilerOptions);
  tsHost.writeFile = (fileName, declaration) => {
    vfs.set(fileName, declaration);
  };
  const _readFile = tsHost.readFile;
  tsHost.readFile = (filename) => {
    if (vfs.has(filename)) {
      return vfs.get(filename);
    }
    return _readFile(filename);
  };
  const program = ts.createProgram(inputFiles, compilerOptions, tsHost);
  await program.emit();
  const output = {};
  for (const filename of inputFiles) {
    const dtsFilename = filename.replace(/\.(m|c)?(ts|js)$/, ".d.$1ts");
    let contents = vfs.get(dtsFilename) || "";
    if (opts?.addRelativeDeclarationExtensions) {
      const ext = filename.match(/\.(m|c)?(ts|js)$/)?.[0].replace(/ts$/, "js") || ".js";
      const imports = mlly.findStaticImports(contents);
      const exports = mlly.findExports(contents);
      const typeExports = mlly.findTypeExports(contents);
      for (const spec of [...exports, ...typeExports, ...imports]) {
        if (!spec.specifier || !/^\.{1,2}[/\\]/.test(spec.specifier)) {
          continue;
        }
        contents = contents.replace(
          spec.code,
          spec.code.replace(spec.specifier, spec.specifier + ext)
        );
      }
    }
    output[filename] = contents;
  }
  return output;
}

async function mkdist(options = {}) {
  options.rootDir = pathe.resolve(process.cwd(), options.rootDir || ".");
  options.srcDir = pathe.resolve(options.rootDir, options.srcDir || "src");
  options.distDir = pathe.resolve(options.rootDir, options.distDir || "dist");
  if (options.cleanDist !== false) {
    await fse.unlink(options.distDir).catch(() => {
    });
    await fse.emptyDir(options.distDir);
    await fse.mkdirp(options.distDir);
  }
  const { globby } = await import('globby');
  const filePaths = await globby(options.pattern || "**", {
    absolute: false,
    cwd: options.srcDir
  });
  const files = filePaths.map((path) => {
    const sourcePath = pathe.resolve(options.srcDir, path);
    return {
      path,
      srcPath: sourcePath,
      extension: pathe.extname(path),
      getContents: () => fse.readFile(sourcePath, { encoding: "utf8" })
    };
  });
  const { loadFile } = createLoader({
    format: options.format,
    ext: options.ext,
    declaration: options.declaration
  });
  const outputs = [];
  for (const file of files) {
    outputs.push(...await loadFile(file) || []);
  }
  for (const output of outputs.filter((o) => o.extension)) {
    const renamed = pathe.basename(output.path, pathe.extname(output.path)) + output.extension;
    output.path = pathe.join(pathe.dirname(output.path), renamed);
    if (outputs.some((o) => o !== output && o.path === output.path)) {
      output.skip = true;
    }
  }
  const dtsOutputs = outputs.filter((o) => o.declaration && !o.skip);
  if (dtsOutputs.length > 0) {
    const declarations = await getDeclarations(
      new Map(dtsOutputs.map((o) => [o.srcPath, o.contents || ""])),
      {
        addRelativeDeclarationExtensions: options.addRelativeDeclarationExtensions
      }
    );
    for (const output of dtsOutputs) {
      output.contents = declarations[output.srcPath] || "";
    }
  }
  const outPaths = new Set(outputs.map((o) => o.path));
  const resolveId = (from, id = "", resolveExtensions) => {
    if (!id.startsWith(".")) {
      return id;
    }
    for (const extension of resolveExtensions) {
      if (outPaths.has(pathe.join(pathe.dirname(from), id + extension))) {
        return id + extension;
      }
    }
    return id;
  };
  const esmResolveExtensions = [
    "",
    "/index.mjs",
    "/index.js",
    ".mjs",
    ".ts",
    ".js"
  ];
  for (const output of outputs.filter(
    (o) => o.extension === ".mjs" || o.extension === ".js"
  )) {
    output.contents = output.contents.replace(
      /(import|export)(\s+(?:.+|{[\s\w,]+})\s+from\s+["'])(.*)(["'])/g,
      (_, type, head, id, tail) => type + head + resolveId(output.path, id, esmResolveExtensions) + tail
    );
  }
  const cjsResolveExtensions = ["", "/index.cjs", ".cjs"];
  for (const output of outputs.filter((o) => o.extension === ".cjs")) {
    output.contents = output.contents.replace(
      /require\((["'])(.*)(["'])\)/g,
      (_, head, id, tail) => "require(" + head + resolveId(output.path, id, cjsResolveExtensions) + tail + ")"
    );
  }
  const writtenFiles = [];
  await Promise.all(
    outputs.filter((o) => !o.skip).map(async (output) => {
      const outFile = pathe.join(options.distDir, output.path);
      await fse.mkdirp(pathe.dirname(outFile));
      await (output.raw ? copyFileWithStream(output.srcPath, outFile) : fse.writeFile(outFile, output.contents, "utf8"));
      writtenFiles.push(outFile);
    })
  );
  return {
    writtenFiles
  };
}

exports.mkdist = mkdist;
